### DISCUSSION TITLE
If I'm changing the time on an existing alarm, and keeping the same id, do I need to stop and set it, or just set it?

### DISCUSSION THREAD

**Discussion by ApplySci** [gdelataillade/alarm#201] [Q&A] [answered]
As per the title - If I'm changing the time on an existing alarm, and keeping the same id, do I need to stop and set it, or just set it?

It's just that I keep getting a crash on iPhone (works fine on android) when I try to set an alarm with same id and same time as currently exists

**Comment by gdelataillade** [comment]
Hi @ApplySci 

If you set an alarm keeping the same `id`, it will first stop the existing alarm and schedule the new one, even if they are exactly the same.

Can you tell me more about your crash on iOS ? How to reproduce, what version ? Make sure you use `await` when you call `Alarm.set`

**Reply by ApplySci** [reply, marked as answer]
I've found a workaround now, which is to call stopAllAlarms before setting any.

I've got quite a convoluted setup for Mac, which might create a situation that is (1) not a real world case and (2) hard to reproduce: I've got a windows pc running Mac Monterey ios on VMWare, and then I'm simulating an iPhone 14 on that.

I've got a few alarms that get set, and sometimes one or two of the times get updated. To keep it simple, I re-set all the alarms when I get a notification that a time has been updated. If a time has changed, then Alarm.set behaves fine. But if the alarm hasn't changed, then it crashes.

But now I'm awaiting stopAllAlarms before setting any, there's no crash.

I may have an actual iPhone to test on in a couple of weeks, and I'll let you know if I can reproduce this on that.

**Reply by ApplySci** [reply]
thread dump: 

```
hread 47 Crashed::  Dispatch queue: com.gdelataillade.alarm.resourceAccessQueue
0   libobjc.A.dylib               	    0x7ff80002d6cb objc_msgSend + 11
1   libswiftCore.dylib            	    0x7ff80d6a474e Dictionary._Variant.removeValue(forKey:) + 190
2   alarm                         	       0x10529fa3f closure #1 in SwiftAlarmPlugin.stopAlarm(id:cancelNotif:result:) + 783 (SwiftAlarmPlugin.swift:293)
3   alarm                         	       0x105299442 closure #1 in SwiftAlarmPlugin.safeModifyResources(_:) + 50 (SwiftAlarmPlugin.swift:67)
4   alarm                         	       0x105298f88 thunk for @escaping @callee_guaranteed () -> () + 40
5   libdispatch.dylib             	    0x7ff80013b7fb _dispatch_call_block_and_release + 12
6   libdispatch.dylib             	    0x7ff80013ca3a _dispatch_client_callout + 8
7   libdispatch.dylib             	    0x7ff800144346 _dispatch_lane_serial_drain + 949
8   libdispatch.dylib             	    0x7ff800144fe3 _dispatch_lane_invoke + 400
9   libdispatch.dylib             	    0x7ff800151b46 _dispatch_workloop_worker_thread + 870
10  libsystem_pthread.dylib       	    0x7ff833e39fd0 _pthread_wqthread + 326
11  libsystem_pthread.dylib       	    0x7ff833e38f57 start_wqthread + 15
```

**Reply by gdelataillade** [reply]
What happened is that when you run `Alarm.set` without waiting for `stopAllAlarms` to finish, my iOS native code will try to access and modify the same resources at the same time and that causes the crash.

**Reply by ApplySci** [reply]
Sorry, I wasn't clear.

stopAllAlarms was the fix. stopAllAlarms wasn't in the code when the crash was happening.
